api/layer/multipart_upload.go

```diff
@@ -0,0 +1,459 @@
+		IsDir    bool
+		Key      string
+		UploadID string
+		Owner    *owner.ID
+		Created  time.Time
+	}
+)
+
+func (n *layer) UploadPart(ctx context.Context, p *UploadPartParams) (*data.ObjectInfo, error) {
+	if p.PartNumber != 0 {
+		_, err := n.GetUploadInitInfo(ctx, p.Info)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if p.Header == nil {
+		p.Header = make(map[string]string)
+	}
+
+	p.Header[UploadIDAttributeName] = p.Info.UploadID
+	p.Header[UploadPartNumberAttributeName] = strconv.Itoa(p.PartNumber)
+	p.Header[UploadKeyAttributeName] = p.Info.Key
+	p.Header[attrVersionsIgnore] = "true"
+
+	params := &PutObjectParams{
+		Bucket: p.Info.Bkt.Name,
+		Object: createUploadPartName(p.Info.UploadID, p.Info.Key, p.PartNumber),
+		Size:   p.Size,
+		Reader: p.Reader,
+		Header: p.Header,
+	}
+
+	return n.objectPut(ctx, p.Info.Bkt, params)
+}
+
+func (n *layer) CompleteMultipartUpload(ctx context.Context, p *CompleteMultipartParams) (*data.ObjectInfo, error) {
+	var (
+		obj             *data.ObjectInfo
+		uploadCompleted = false
+	)
+
+	for i := 1; i < len(p.Parts); i++ {
+		if p.Parts[i].PartNumber < p.Parts[i-1].PartNumber {
+			return nil, errors.GetAPIError(errors.ErrInvalidPartOrder)
+		}
+	}
+
+	objects, err := n.getUploadParts(ctx, p.Info.UploadID, p.Info.Key, p.Info.Bkt)
+	if err != nil {
+		if err == errors.GetAPIError(errors.ErrNoSuchUpload) {
+			return n.headLastVersionIfNotDeleted(ctx, p.Info.Bkt, p.Info.Key)
+		}
+		return nil, err
+
+	}
+
+	if len(objects) < len(p.Parts) {
+		return nil, errors.GetAPIError(errors.ErrInvalidPart)
+	}
+
+	parts := make([]*data.ObjectInfo, 0, len(p.Parts))
+
+	for i, part := range p.Parts {
+		info := objects[part.PartNumber]
+		if info == nil || part.ETag != info.HashSum {
+			return nil, errors.GetAPIError(errors.ErrInvalidPart)
+		}
+		if i != len(p.Parts)-1 && info.Size <= UploadMinSize {
+			return nil, errors.GetAPIError(errors.ErrEntityTooSmall)
+		}
+		parts = append(parts, info)
+	}
+
+	initMetadata := objects[0].Headers
+
+	pr, pw := io.Pipe()
+
+	done := make(chan bool)
+
+	delete(initMetadata, UploadIDAttributeName)
+	delete(initMetadata, UploadPartNumberAttributeName)
+	delete(initMetadata, UploadKeyAttributeName)
+	delete(initMetadata, attrVersionsIgnore)
+
+	go func(done chan bool) {
+		obj, err = n.objectPut(ctx, p.Info.Bkt, &PutObjectParams{
+			Bucket: p.Info.Bkt.Name,
+			Object: p.Info.Key,
+			Reader: pr,
+			Header: initMetadata,
+		})
+		if err != nil {
+			done <- true
+			return
+		}
+		uploadCompleted = true
+		done <- true
+	}(done)
+	for _, part := range parts {
+		_, err := n.objectGetWithPayloadWriter(ctx, &getParams{
+			Writer: pw,
+			cid:    p.Info.Bkt.CID,
+			oid:    part.ID,
+		})
+		if err != nil {
+			n.log.Error("could not download a part of multipart upload",
+				zap.String("uploadID", p.Info.UploadID),
+				zap.String("part number", part.Headers[UploadPartNumberAttributeName]),
+				zap.Error(err))
+			return nil, err
+		}
+	}
+	pw.Close()
```

Comment:
I suppose we should handle the error

Reply:
I'm not sure, because, if I understand correctly,  `Close()` always returns `nil`: 
```
func (w *PipeWriter) Close() error {
	return w.CloseWithError(nil)
}

func (w *PipeWriter) CloseWithError(err error) error {
	return w.p.CloseWrite(err)
}

func (p *pipe) CloseWrite(err error) error {
	if err == nil {
		err = EOF
	}
	p.werr.Store(err)
	p.once.Do(func() { close(p.done) })
	return nil
}
```

Reply:
Then let's explicitly ignore the error.
```suggestion
	_ = pw.Close()
```